pkg/user/persister/persister.go

```diff
@@ -213,42 +213,58 @@ func (p *persister) runWorker() error {
 }
 
 func (p *persister) handleChunk(chunk map[string]*puller.Message) {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	events := make(map[string][]*eventproto.UserEvent)
+	messages := make(map[string][]*puller.Message)
 	for _, msg := range chunk {
-		event, err := p.unmarshalMessage(msg)
-		if err != nil {
-			handledCounter.WithLabelValues(codes.BadMessage.String()).Inc()
+		// Extract the user event
+		event := p.extractUserEvent(msg)
+		if event == nil {
 			continue
 		}
-		if !p.validateEvent(event) {
-			handledCounter.WithLabelValues(codes.BadMessage.String()).Inc()
-			continue
+		// Append events per environment
+		listEvents, ok := events[event.EnvironmentNamespace]
+		if ok {
+			events[event.EnvironmentNamespace] = append(listEvents, event)
+		} else {
+			events[event.EnvironmentNamespace] = []*eventproto.UserEvent{event}
+		}
+		// Append PubSub messages per environment
+		listMessages, ok := messages[event.EnvironmentNamespace]
+		if ok {
+			messages[event.EnvironmentNamespace] = append(listMessages, msg)
+		} else {
+			messages[event.EnvironmentNamespace] = []*puller.Message{msg}
 		}
-		ok, repeatable := p.upsert(event)
-		if !ok {
-			if repeatable {
-				msg.Nack()
-				handledCounter.WithLabelValues(codes.RepeatableError.String()).Inc()
+	}
+	// Upsert events
+	if len(events) > 0 {
+		for environmentNamespace, events := range events {
+			// Upsert events per environment
+			err := p.upsertMAUs(ctx, events, environmentNamespace)
+			if err != nil {
```

Comment:
I think we can log the error here for debugging purposes.

Reply:
The `upsertMAUs` function is printing the logs.

https://github.com/bucketeer-io/bucketeer/pull/936/files#diff-7e188d43cb88893445af88b53e185b56a395b08fccc3f55e82e1bb01c6e5b940R324-R330